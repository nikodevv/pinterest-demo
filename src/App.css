/* Commonly used containers */
.container-row {
  display: flex;
  flex-direction: row;
  width: 100%;
}

.container-column {
  display: flex;
  flex-direction: column;
  width:100%;
}

.center {
  justify-content: center;
}

/* Overlays entire screen in with a low opacity background. Used for all modals*/
.screenOverlay {
  position: fixed;
  background-color: rgba(54,54,84,0.50);
  height: 100vh;
  width: 100vw;
  z-index: 5000;

  display: flex;
  justify-content: center;
  align-items: center;
}
