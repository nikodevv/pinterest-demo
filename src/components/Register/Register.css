.formContainer {
  display: flex;
  width: 40%;
  height: 30%;
  border-radius: 0.5rem;
  flex-direction: column;
  align-items: center;
  justify-content: center;
  background-color: white;
}

.formContainer input {
  outline: 0;
  width: 70%;
  height: 1.9rem;
  border-radius: 0.25rem;
  border-color: transparent;
  box-shadow: 0 0 0 0.1rem #9d9d9d;
  margin-top: 2rem;
}
.formContainer input:focus {
  box-shadow: 0 0 0 0.25rem #91b9ff;
}
