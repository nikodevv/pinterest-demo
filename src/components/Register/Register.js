import React, { Component } from 'react';
import {useSelector} from "react-redux";

export const Register = () => {
  const username = useSelector( state => state.auth.username);
  if (username === null) {
    return null;
  }
  return (
    <div>full page</div>
  )
}

export default Register;
