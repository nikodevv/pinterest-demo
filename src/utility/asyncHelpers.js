/* *
 * A set of pure functions that help components perform async actions.
 */
