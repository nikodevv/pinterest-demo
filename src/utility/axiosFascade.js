import firebaseConfig from '../config/firebaseConfig';
import axios from "axios";

const API_KEY = firebaseConfig.apiKey;

export class AxiosWrapper {
  static shortenLink = (link) => {
    let linkRequest = {
      destination: link,
      domain: { fullName: "rebrand.ly" }
    };
    let requestHeaders = {
      "Content-Type": "application/json",
      "apikey": "d48c10a11e0d477792cdb4de04cea20e",
      "workspace": "2bdcf71bcb2348e1b511aad23b4999dd"
    }
    return axios.post(
      "https://api.rebrandly.com/v1/links",
      linkRequest,
      {headers: requestHeaders}
      )
  }
}
