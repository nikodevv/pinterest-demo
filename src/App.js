import React from 'react';
import * as firebase from 'firebase/app';
import "firebase/auth";
import "firebase/firestore";
import { Provider } from 'react-redux';
import { createStore, applyMiddleware } from "redux";
import thunk from 'redux-thunk';
import './App.css';
import firebaseConfig from './config/firebaseConfig';
import Router from "./components/Router";
import rootReducer from './reducers'

const firebaseApp = firebase.initializeApp(firebaseConfig);
export const db = firebase.firestore(firebaseApp);

const store = createStore(
    rootReducer,
    applyMiddleware(thunk)
);

const App = () => {
  return (
      <Provider store={store}>
        <div className="App">
          <Router/>
        </div>
      </Provider>
  );
};

export default App;
