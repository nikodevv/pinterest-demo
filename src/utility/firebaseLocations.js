/*
* Firebase collection references are separated from other code for easy mocking via dependency injection.
* */
import { db } from '../App';

export let UsersRef = () => db.collection('users');

