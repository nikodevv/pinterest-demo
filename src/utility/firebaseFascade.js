import firebase from "firebase";
import {UsersRef} from './firebaseLocations';

export class FirebaseAuth {
  static loginWithGithub = () => {
    const provider = new firebase.auth.GithubAuthProvider();
    return firebase.auth().signInWithPopup(provider)
  };

  static signOut = () => {
    return firebase.auth().signOut()
  };
}

export class FirestoreData {
  static addPost = (post) => {
    const uid = firebase.auth().currentUser.uid;
    return UsersRef().doc(uid).collection('posts').add(post);
  };

  static findUsersWithUsername = async (queryStr) => {
    const query = await UsersRef().where('username','>=', queryStr);
    const snapshot = await query.get();
    const userModels = [];
    snapshot.forEach(doc => {
      const userObj = {...doc.data(), id: doc.id };
      userModels.push(userObj)
    });
    return userModels;
  };

  static setUsername = (username) => {
    const uid = firebase.auth().currentUser.uid;
    return UsersRef().doc(uid).set({username}, {merge: true});
  };
  static fetchUserPosts = async (uid) => {
    const query = await UsersRef().doc(uid).collection('posts').get();
    if (query.empty === true){
      return []
    }
    const posts = [];
    query.forEach(doc=> {
      posts.push({
        id: doc.id,
        ...doc.data()
      })
    });
    return posts
  };

  static fetchOwnUserModel = async () => {
    const uid = firebase.auth().currentUser.uid;
    const snapshot = await UsersRef().doc(uid).get();
    if (snapshot.exists && snapshot.data().username !== undefined) {
      return snapshot.data();
    }
    return {
      username: null
    }
  };

  static deletePost = (uid, itemId) => {
    return UsersRef().doc(uid).collection('posts').doc(itemId).delete();
  }
}
