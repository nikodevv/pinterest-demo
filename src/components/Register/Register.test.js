import React from 'react';
import {configure, mount, shallow} from 'enzyme';
import configureMockStore from 'redux-mock-store';
import Adapter from "enzyme-adapter-react-16/build";
import {Provider} from "react-redux";
import Register, {helpers} from './Register';
import {firebaseAuth} from "../../utility/firebaseFascade";
import {App} from '../../App';

configure({adapter: new Adapter()});
const mockStore = configureMockStore();
window.alert = jest.fn(); // must be defined in outer scope
describe('<Register />', () => {
  let initialStoreData;
  let event = {
    preventDefault: ()=>null
  };
  beforeEach(()=>{
    initialStoreData = {
      auth: {
        loading: false,
        loggedIn: true,
        username: 'validUsername',
      }
    };
    firebaseAuth.findUsersWithUsername = jest.fn(()=>[]);
    firebaseAuth.setUsername = jest.fn();
    window.alert = jest.fn();
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  test('renders', () => {
    const wrapper = shallow(
      <Provider store={mockStore(initialStoreData)}>
        <Register/>
      </Provider>);
    expect(wrapper).toMatchSnapshot();
  });

  test('renders null if auth is loading', () => {
    initialStoreData.auth.loading = true;
    const wrapper = mount(
      <Provider store={mockStore(initialStoreData)}>
        <Register/>
      </Provider>);
    expect(wrapper.find('div').isEmpty()).toEqual(true);
  });

  test('renders null if username is not null and loggedIn is true', () => {
    const wrapper = mount(
      <Provider store={mockStore(initialStoreData)}>
        <Register/>
      </Provider>);
    expect(wrapper.find('div').isEmpty()).toEqual(true);
  });

  test('renders registration form if username is null and logged in is true', () => {
    initialStoreData.auth.username = null;
    const wrapper = mount(
      <Provider store={mockStore(initialStoreData)}>
        <Register/>
      </Provider>);
    expect(wrapper.find('form').isEmpty()).toEqual(false);
  });

  test('alphanumeric filter strips out non alphanumeric chars', () => {
    const testValue = 'Test123*** `&),{}TEST';
    const val = helpers.filterAlphanumerics(testValue);
    expect(val).toEqual('Test123TEST');
  });

  test('setUsername is called on input value change', () => {
    initialStoreData.auth.username = null;
    const testInput1 = 'iceFlag'.split('');
    const setUsername = jest.fn();
    const mockUseState = jest.fn((unusedValue) => {
      return [unusedValue, setUsername]
    });
    jest.spyOn(React, 'useState').mockImplementation(mockUseState);
    helpers.filterAlphanumerics = (x)=>x;
    expect(mockUseState).toBeCalledTimes(0);
    const wrapper = mount(
      <Provider store={mockStore(initialStoreData)}>
        <Register/>
      </Provider>);
    expect(mockUseState).toBeCalledTimes(1);
    expect(wrapper.find('#usernameInput').isEmpty()).toEqual(false);
    testInput1.forEach((key) => {
      wrapper.find('#usernameInput').simulate('change', {target:{value: key}});
      expect(setUsername).toBeCalledWith(key);
    })
  });

  test('submit form handler calls firebase Fascade endpoint to check if username exists', ()=> {
    expect(firebaseAuth.findUsersWithUsername).toBeCalledTimes(0);
    helpers.handleSubmit('username', jest.fn(), event);
    expect(firebaseAuth.findUsersWithUsername).toBeCalledTimes(1);
  });

  test('submit form handler returns alert if username exists in database, does not call setUsername', async () => {
    const username = 'myusername';
    firebaseAuth.findUsersWithUsername.mockImplementation(()=>([{username}]));
    expect(window.alert).toBeCalledTimes(0);
    await helpers.handleSubmit(username, jest.fn(), event);
    expect(window.alert).toBeCalledTimes(1);
    expect(firebaseAuth.setUsername).toBeCalledTimes(0);
  });

  test('submit form handler calls set username in fascade if username is not taken', async ()=>{
    const username = 'myusername';
    expect(window.alert).toBeCalledTimes(0);
    await helpers.handleSubmit(username, jest.fn(), event);
    expect(window.alert).toBeCalledTimes(0);
    expect(firebaseAuth.setUsername).toBeCalledTimes(1);
    expect(firebaseAuth.setUsername).toBeCalledWith(username);
  });

  test('if form handler setusername promise is resolved, calls dispatch', async ()=> {
    const username = 'myusername';
    const mockDispatch = jest.fn();
    expect(firebaseAuth.setUsername).toBeCalledTimes(0);
    expect(mockDispatch).toBeCalledTimes(0);
    await helpers.handleSubmit(username, mockDispatch, event);
    expect(firebaseAuth.setUsername).toBeCalledTimes(1);
    expect(mockDispatch).toBeCalledTimes(1);
  });

  test('form handler stop propagation', async () => {
    const username = 'myusername';
    event.preventDefault = jest.fn();
    expect(event.preventDefault).toBeCalledTimes(0);
    await helpers.handleSubmit(username, jest.fn(), event);
    expect(event.preventDefault).toBeCalledTimes(1);
  })
});
