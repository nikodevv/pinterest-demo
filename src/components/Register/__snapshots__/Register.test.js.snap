// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Register /> renders 1`] = `ShallowWrapper {}`;
